crates/dht/src/dht.rs

```diff
@@ -0,0 +1,447 @@
+    ///
+    /// The state flow is as follows:
+    /// - If the two snapshots are identical, the function will return [DhtSnapshotNextAction::Identical].
+    /// - If the two snapshots cannot be compared, the function will return [DhtSnapshotNextAction::CannotCompare].
+    ///   This can happen if the time slices of the two DHTs are not aligned or one side is following
+    ///   a different flow to what we're expecting.
+    /// - If the snapshots are different, the function will return [DhtSnapshotNextAction::NewSnapshot]
+    ///   with a more detailed snapshot of the DHT model.
+    /// - When the most detailed snapshot type is reached, the function will return [DhtSnapshotNextAction::NewSnapshotAndHashList]
+    /// - The new snapshot from [DhtSnapshotNextAction::NewSnapshotAndHashList] should be sent to the
+    ///   other party so that they can compare it with their own snapshot and determine which op hashes
+    ///   they need to fetch.
+    /// - On the final comparison step, the function will return [DhtSnapshotNextAction::HashList]
+    ///   with a list of op hashes. This list should be sent to the other party so that they can fetch
+    ///   any missing ops.
+    ///
+    /// Notice that the final step would require re-computing the most detailed snapshot type. This
+    /// is expensive. To avoid having to recompute a snapshot we've already computed, the caller
+    /// MUST capture the snapshot from [DhtSnapshotNextAction::NewSnapshot] when it contains either
+    /// [DhtSnapshot::DiscSectorDetails] or [DhtSnapshot::RingSectorDetails]. This snapshot can be
+    /// provided back to this function in the `our_previous_snapshot` parameter. In all other cases,
+    /// the caller should provide `None` for `our_previous_snapshot`.
+    ///
+    /// Note also that there are two possible routes through the comparison process. The first is
+    /// when the historical disc mismatches, the second is when the recent rings mismatch. The
+    /// historical disc mismatch is prioritised, so if a mismatch is detected there then the sync
+    /// process will resolve that. Otherwise, the recent rings mismatch will be resolved. That means
+    /// that it may take up to two rounds of sync to resolve all mismatches. Of course, both the
+    /// disc and the rings must be considered a moving target so it cannot be assumed that 2 rounds
+    /// are actually enough to resolve all mismatches.
+    ///
+    /// The `arc_set` parameter is used to determine which arcs are relevant to the DHT model. This
+    /// should be the [ArcSet::intersection] of the arc sets of the two DHT models to be compared.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if there are no arcs to snapshot. If there is no overlap between the arc
+    /// sets of two DHT models then there is no point in comparing them because it will always
+    /// yield an empty diff. The [ArcSet::covered_sector_count] should be checked before calling
```

Comment:
I guess that's the caller's choice. I'd prefer to avoid an error that I can only convert to a string to tell it apart from other errors. Or even with strong types, if I know an input won't work for a function, I'd prefer to filter it out.

Gossip should be choosing other nodes with agent arcs that overlap with ours and the node accepting a gossip round should be checking what the initiator thought was the arc overlap so there's really no reason for an empty ArcSet to get this far. I added the error just to be doubly sure we find out about mistakes elsewhere because it'll be across a crate boundary :)